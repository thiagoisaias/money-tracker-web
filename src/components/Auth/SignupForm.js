import React, { Component, Fragment } from "react";
import styled from "styled-components";

const Form = styled.form`
  max-width: 500px;
  margin: 0 auto;
`;

const Input = styled.input`
  display: block;
  height: 40px;
  width: calc(100% - 12px);
  border: 1px solid #eee;
  border-radius: 3px;
  font-size: 14px;
  margin: 6px 0;
  padding-left: 12px;

  &:focus {
    border: 1px solid #d2d2d2;
  }
`;

const SubmitButton = styled.button`
  width: 100%;
  height: 40px;
  font-size: 12px;
  background-color: #333;
  cursor: pointer;
  border-radius: 3px;
  color: #fff;
  margin-top: 32px;

  &:hover {
    opacity: 0.9;
  }
`;

const Title = styled.h2`
  text-align: center;
  font-size: 1.2em;
  font-weight: 600;
  margin-bottom: 32px;
`;

const Alternate = styled.div`
  font-size: 12px;
  color: #333;
  margin-top: 32px;
  text-align: center;
  > strong {
    cursor: pointer;
    font-weight: 600;
  }
`;

class SignupForm extends Component {
  constructor(props) {
    super(props);
    this.state = {
      email: "",
      password: "",
      passwordConfirmation: ""
    };

    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  handleChange(event) {
    const target = event.target;
    const value = target.value;
    const name = target.name;

    this.setState({
      [name]: value
    });
  }

  handleSubmit(event) {
    event.preventDefault();
  }

  render() {
    const { toggleNewAccount } = this.props;
    return (
      <Fragment>
        <Title>Create Account</Title>
        <Form onSubmit={this.handleSubmit}>
          <Input
            type="text"
            name="email"
            placeholder="Email"
            value={this.state.email}
            onChange={this.handleChange}
          />
          <Input
            type="password"
            name="password"
            placeholder="Password"
            value={this.state.password}
            onChange={this.handleChange}
          />
          <Input
            type="password"
            name="passwordConfirmation"
            placeholder="Confirm your password"
            value={this.state.passwordConfirmation}
            onChange={this.handleChange}
          />
          <SubmitButton onClick={this.handleLogin}> Sign Up </SubmitButton>
        </Form>
        <Alternate>
          Already have an account?{" "}
          <strong onClick={toggleNewAccount}>Log in here </strong>
        </Alternate>
      </Fragment>
    );
  }
}

export default SignupForm;
