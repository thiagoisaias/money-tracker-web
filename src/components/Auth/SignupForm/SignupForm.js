import React, { Component, Fragment } from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { NavLink } from "react-router-dom";

import Spinner from "../../UI/Spinner/Spinner";
import withFormInputHandler from "../../../hoc/withFormInputHandler/withFormInputHandler";

const ErrorMessage = styled.p`
  color: salmon;
  font-size: 14px;
  margin: 16px 0;
`;

const Form = styled.form`
  width: 100%;
`;

const Input = styled.input`
  display: block;
  height: 40px;
  width: calc(100% - 18px);
  border: 1px solid #eee;
  border-radius: 2px;
  font-family: inherit;
  color: inherit;
  font-size: 14px;
  margin: 6px 0;
  padding-left: 16px;

  &:focus {
    border: 1px solid #d2d2d2;
  }

  &::placeholder {
    opacity: 0.7;
  }
`;

const SubmitButton = styled.button`
  width: 100%;
  height: 40px;
  font-size: 12px;
  background-color: #333;
  cursor: pointer;
  border-radius: 2px;
  color: #fff;
  margin-top: 16px;
`;

const Title = styled.h2`
  text-align: center;
  font-weight: 600;
  margin-bottom: 32px;
  font-size: 22px;
`;

const Alternate = styled.div`
  font-size: 12px;
  color: #333;
  margin-top: 32px;
  text-align: center;
`;

const Link = styled(NavLink)`
  cursor: pointer;
  font-weight: 600;
`;

export class SignupForm extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isFormValid: false,
      formFields: {
        name: {
          elementType: "input",
          elementConfig: {
            type: "text",
            placeholder: "Full name",
            autoComplete: "name"
          },
          name: "Name",
          value: "",
          validation: {
            required: true
          },
          isValid: false,
          touched: false
        },
        email: {
          elementType: "input",
          elementConfig: {
            type: "text",
            placeholder: "Email",
            autoComplete: "email"
          },
          name: "Email",
          value: "",
          validation: {
            required: true
          },
          isValid: false,
          touched: false
        },
        password: {
          elementType: "input",
          elementConfig: {
            type: "password",
            placeholder: "Password",
            autoComplete: "password"
          },
          name: "Password",
          value: "",
          validation: {
            required: true
          },
          isValid: false,
          touched: false
        },
        passwordConfirmation: {
          elementType: "input",
          elementConfig: {
            type: "password",
            placeholder: "Password confirmation",
            autoComplete: "passwordConfirmation"
          },
          name: "Password",
          value: "",
          validation: {
            required: true
          },
          isValid: false,
          touched: false
        }
      }
    };
  }

  checkFieldValidity = (value, rules) => {
    let isValid = true;

    if (rules.required) {
      isValid = value.trim() !== "" && isValid;
    }

    return isValid;
  };

  onInputChange = event => {
    const { handleInputChange } = this.props;
    const formFields = { ...this.state.formFields };
    const checkFieldValidity = this.checkFieldValidity;
    const updatedFormState = handleInputChange(
      event,
      formFields,
      checkFieldValidity
    );
    this.setState({
      ...updatedFormState
    });
  };

  handleSubmit = event => {
    const { submitData } = this.props;
    event.preventDefault();

    const formData = {};

    for (let fieldKey in this.state.formFields) {
      formData[fieldKey] = this.state.formFields[fieldKey].value;
    }

    submitData(formData);
  };

  render() {
    const { isLoading, error } = this.props;

    const keyList = [];
    for (let key in this.state.formFields) {
      keyList.push(key);
    }

    const inputList = keyList.map(key => {
      const formField = this.state.formFields[key];
      return (
        <Input
          key={key}
          {...formField.elementConfig}
          name={key}
          value={formField.value}
          touched={formField.touched}
          isValid={formField.isValid}
          onChange={this.onInputChange}
        />
      );
    });

    return (
      <Fragment>
        <Title>{"Create an account"}</Title>
        <Form onSubmit={this.handleSubmit}>
          {error && <ErrorMessage> {error} </ErrorMessage>}
          {inputList}
          <SubmitButton onClick={this.handleSubmit}>
            {isLoading ? <Spinner width={35} height={35} /> : "Sign Up"}
          </SubmitButton>
        </Form>
        <Alternate>
          {"Already have an account? "}
          <Link to="/auth/login">{"Login here"}</Link>
        </Alternate>
      </Fragment>
    );
  }
}

SignupForm.propTypes = {
  error: PropTypes.string,
  handleInputChange: PropTypes.func.isRequired,
  isLoading: PropTypes.bool.isRequired,
  submitData: PropTypes.func.isRequired
};

export default withFormInputHandler(SignupForm);
