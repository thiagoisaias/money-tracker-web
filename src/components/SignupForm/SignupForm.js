import React, { Component, Fragment } from "react";
import styled from "styled-components";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import Spinner from "../Spinner/Spinner";
import { Redirect } from "react-router-dom";
import { signup } from "../../store/actions/auth";

const ErrorMessage = styled.p`
  color: #e75252;
  font-size: 14px;
  margin: 12px 1px;
`;

const Form = styled.form`
  max-width: 500px;
  margin: 0 auto;
`;

const Input = styled.input`
  display: block;
  height: 40px;
  width: calc(100% - 12px);
  border: 1px solid #eee;
  border-radius: 3px;
  font-size: 14px;
  margin: 6px 0;
  padding-left: 12px;

  &:focus {
    border: 1px solid #d2d2d2;
  }
`;

const SubmitButton = styled.button`
  width: 100%;
  height: 40px;
  font-size: 12px;
  background-color: #333;
  cursor: pointer;
  border-radius: 3px;
  color: #fff;
  margin-top: 32px;
`;

const Title = styled.h2`
  text-align: center;
  font-weight: 600;
  margin-bottom: 32px;
  font-size: 22px;
`;

const Alternate = styled.div`
  font-size: 12px;
  color: #333;
  margin-top: 32px;
  text-align: center;
  > strong {
    cursor: pointer;
    font-weight: 600;
  }
`;

export class SignupForm extends Component {
  constructor(props) {
    super(props);
    this.state = {
      name: "",
      email: "",
      password: "",
      passwordConfirmation: ""
    };
  }

  handleChange = event => {
    const target = event.target;
    const value = target.value;
    const name = target.name;

    this.setState({
      [name]: value
    });
  };

  handleSubmit = event => {
    event.preventDefault();
    if (this.props.isLoading) {
      return;
    }
    this.props.onSignup(this.state);
  };

  render() {
    const { toggleNewAccount, isAuthenticated, isLoading, error } = this.props;
    const authRedirect = <Redirect to="/" />;

    return (
      <Fragment>
        {isAuthenticated ? authRedirect : null}
        <Title>Create Account</Title>
        {error && <ErrorMessage> {error.full_messages[0]} </ErrorMessage>}
        <Form onSubmit={this.handleSubmit}>
          <Input
            type="text"
            name="name"
            placeholder="Name"
            value={this.state.name}
            onChange={this.handleChange}
          />
          <Input
            type="text"
            name="email"
            placeholder="Email"
            value={this.state.email}
            onChange={this.handleChange}
          />
          <Input
            type="password"
            name="password"
            placeholder="Password"
            value={this.state.password}
            onChange={this.handleChange}
          />
          <Input
            type="password"
            name="passwordConfirmation"
            placeholder="Confirm your password"
            value={this.state.passwordConfirmation}
            onChange={this.handleChange}
          />
          <SubmitButton onClick={this.handleSubmit}>
            {isLoading ? <Spinner width={35} height={35} /> : "Sign Up"}
          </SubmitButton>
        </Form>
        <Alternate>
          Already have an account?{" "}
          <strong onClick={toggleNewAccount}>Log in here </strong>
        </Alternate>
      </Fragment>
    );
  }
}

SignupForm.propTypes = {
  toggleNewAccount: PropTypes.func.isRequired,
  onSignup: PropTypes.func.isRequired,
  isAuthenticated: PropTypes.bool.isRequired,
  isLoading: PropTypes.bool.isRequired,
  error: PropTypes.object
};

const mapStateToProps = state => {
  return {
    isAuthenticated: state.auth.isAuthenticated,
    isLoading: state.auth.isLoading,
    error: state.auth.error
  };
};

const mapDispatchToProps = dispatch => {
  return {
    onSignup: signupData => {
      dispatch(signup(signupData));
    }
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(SignupForm);
