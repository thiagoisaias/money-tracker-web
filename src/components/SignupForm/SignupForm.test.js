import React from "react";
import ReactDOM from "react-dom";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import { SignupForm } from "./SignupForm";

describe("SignupForm", () => {
  const mockProps = {
    toggleNewAccount: jest.fn(),
    onSignup: jest.fn(),
    isAuthenticated: true,
    isLoading: false,
    error: null
  };

  const component = shallow(<SignupForm {...mockProps} />);

  it("renders properly", () => {
    expect(toJson(component)).toMatchSnapshot();
  });
});
