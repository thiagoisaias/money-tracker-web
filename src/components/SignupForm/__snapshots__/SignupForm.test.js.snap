// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignupForm renders properly 1`] = `
<React.Fragment>
  <Redirect
    push={false}
    to="/"
  />
  <styled.h2>
    Create Account
  </styled.h2>
  <styled.form
    onSubmit={[Function]}
  >
    <styled.input
      name="name"
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value=""
    />
    <styled.input
      name="email"
      onChange={[Function]}
      placeholder="Email"
      type="text"
      value=""
    />
    <styled.input
      name="password"
      onChange={[Function]}
      placeholder="Password"
      type="password"
      value=""
    />
    <styled.input
      name="passwordConfirmation"
      onChange={[Function]}
      placeholder="Confirm your password"
      type="password"
      value=""
    />
    <styled.button
      onClick={[Function]}
    >
      Sign Up
    </styled.button>
  </styled.form>
  <styled.div>
    Already have an account?
     
    <strong
      onClick={[Function]}
    >
      Log in here 
    </strong>
  </styled.div>
</React.Fragment>
`;
